File: src/runtime/boxes/BbCapture.java
package snobol4.runtime.boxes;

/**
 * BbCapture.java — CAPTURE: wrap child; $ writes on every γ; . buffers for Phase-5 commit
 * Port of bb_capture.c / bb_capture.s
 *
 *   CAP_α:       cr = child(α);  if empty → CAP_ω;  else → CAP_γ_core;
 *   CAP_β:       cr = child(β);  if empty → CAP_ω;  else → CAP_γ_core;
 *   CAP_γ_core:  if (immediate) NV_SET(varname, cr);
 *                else           pending=cr; has_pending=true;
 *                               return cr;
 *   CAP_ω:       has_pending=false; return spec_empty;
 *
 * The executor calls commitPending() after overall match success to flush
 * deferred (.) captures — mirroring match.clj commit-pending! and stmt_exec.c.
 */
public class BbCapture extends BbBox {

    /** Callback interface for variable assignment (NV_SET_fn equivalent) */
    public interface VarSetter {
        void set(String varname, String value);
    }

    private final BbBox     child;
    private final String    varname;
    private final boolean   immediate;   /* true = $var; false = .var */
    private final VarSetter setter;

    /* deferred capture state (.var path) */
    private Spec    pending    = null;
    private boolean hasPending = false;

    public BbCapture(MatchState ms, BbBox child, String varname,
                     boolean immediate, VarSetter setter) {
        super(ms);
        this.child     = child;
        this.varname   = varname;
        this.immediate = immediate;
        this.setter    = setter;
    }

    @Override public Spec alpha() { return runChild(child.alpha()); }
    @Override public Spec beta()  { return runChild(child.beta());  }

    private Spec runChild(Spec cr) {
        if (cr == null) {
            // CAP_ω
            hasPending = false;
            return null;
        }
        // CAP_γ_core
        if (varname != null && !varname.isEmpty()) {
            String matched = ms.sigma.substring(cr.start, cr.start + cr.len);
            if (immediate) {
                setter.set(varname, matched);                                  // $ — write now
            } else {
                pending    = cr;                                               // . — buffer
                hasPending = true;
            }
        }
        return cr;
    }

    /**
     * Called by executor after overall match succeeds to commit deferred captures.
     * Mirrors match.clj commit-pending! and stmt_exec.c Phase-5 flush.
     */
    public void commitPending() {
        if (hasPending && pending != null && varname != null && !varname.isEmpty()) {
            setter.set(varname, ms.sigma.substring(pending.start, pending.start + pending.len));
            hasPending = false;
            pending    = null;
        }
    }
}

File: src/runtime/boxes/capture/bb_capture.c
/* _XNME/_XFNME  CAPTURE     $ writes on every γ; . buffers for Phase-5 commit */
#include "bb_box.h"
#include "../../snobol4/snobol4.h"
#include <stdlib.h>
#include <string.h>

typedef struct {
    bb_box_fn fn; void *state; const char *varname;
    int immediate; spec_t pending; int has_pending;
} capture_t;

spec_t bb_capture(void *zeta, int entry)
{
    capture_t *ζ = zeta;
    spec_t cr;
    if (entry==α)                                                               goto CAP_α;
    if (entry==β)                                                               goto CAP_β;
    CAP_α:          cr=ζ->fn(ζ->state,α);                                       
                    if (spec_is_empty(cr))                                      goto CAP_ω;
                                                                                goto CAP_γ_core;
    CAP_β:          cr=ζ->fn(ζ->state,β);                                       
                    if (spec_is_empty(cr))                                      goto CAP_ω;
                                                                                goto CAP_γ_core;
    CAP_γ_core:     if (ζ->varname && *ζ->varname && ζ->immediate) {            
                        char *s=GC_MALLOC(cr.δ+1);                              
                        memcpy(s,cr.σ,(size_t)cr.δ); s[cr.δ]=0;                 
                        DESCR_t v={.v=DT_S,.slen=(uint32_t)cr.δ,.s=s};          
                        NV_SET_fn(ζ->varname,v);                                
                    } else if (ζ->varname && *ζ->varname) {                     
                        ζ->pending=cr; ζ->has_pending=1; }                      
                                                                                return cr;
    CAP_ω:          ζ->has_pending=0;                                           return spec_empty;
}

capture_t *bb_capture_new(bb_box_fn fn, void *state, const char *varname, int immediate)
{ capture_t *ζ=calloc(1,sizeof(capture_t)); ζ->fn=fn; ζ->state=state; ζ->varname=varname; ζ->immediate=immediate; return ζ; }
